File: String/38_CountAndSay.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
    """ Quite straight-forward solution.

    We generate k-th string, and from k-th string we generate k+1-th string,
    until we generate n-th string.
    """
    def countAndSay(self, n):
        if n <= 1:
            return "1"

        pre_str = "1"
        for i in range(2, n + 1):
            # Get the ith count-and-say sequence by scan pre_str
            length = len(pre_str)
            current_str = ""

            # Count and say the pre_str
            index = 0
            while index < length:
                char = pre_str[index]
                repeat = 0
                pos = index + 1
                while pos < length and pre_str[pos] == char:
                    repeat += 1
                    pos += 1

                current_str += str(repeat + 1) + char
                index = pos

            pre_str = current_str

        return pre_str

"""
1
5
15
"""

File: String/38_CountAndSay.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-07-14 19:12:37
 */

class Solution {
public:
    string countAndSay(int n) {
        int index = 2;
        string pre_str = "1";
        while(index <= n){
            string new_str = "";
            int i=0;
            while(i < pre_str.size()){
                int count = 1;
                char c = pre_str[i];
                i += 1;
                while(i < pre_str.size() && pre_str[i]==c){
                    count++;
                    i++;
                }
                new_str += to_string(count) + c;
            }
            pre_str = new_str;
            index ++;
        }
        return pre_str;
    }
};

/*
1
5
15
*/
